language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/CodeGenRunner.scala

```diff
@@ -109,6 +109,78 @@ object CodeGenRunner extends StrictLogging {
     logger.warn(s"Finish writing file '$outputFile'")
   }
 
+  /** Given the package prefixes specified per DAR and the module-prefixes specified in
+    * daml.yaml, produce the combined prefixes per package id.
+    */
+  def resolvePackagePrefixes(
+      pkgPrefixes: Map[PackageId, String],
+      modulePrefixes: Map[PackageReference, String],
+      interfaces: Seq[Interface],
+  ): Map[PackageId, String] = {
+    val metadata: Map[PackageReference.NameVersion, PackageId] = interfaces.view
+      .flatMap(iface =>
+        iface.metadata.iterator.map(metadata =>
+          PackageReference.NameVersion(metadata.name, metadata.version) -> iface.packageId
+        )
+      )
+      .toMap
+    def resolveRef(ref: PackageReference): PackageId = ref match {
+      case nameVersion: PackageReference.NameVersion =>
+        metadata.getOrElse(
+          nameVersion,
+          throw new IllegalArgumentException(
+            s"""No package $nameVersion found, available packages: ${metadata.keys.mkString(
+              ", "
+            )}"""
+          ),
+        )
+    }
+    val resolvedModulePrefixes: Map[PackageId, String] = modulePrefixes.view.map { case (k, v) =>
+      resolveRef(k) -> v.toLowerCase
+    }.toMap
+    val resolvedPrefixes: Map[PackageId, String] =
+      (pkgPrefixes.keySet union resolvedModulePrefixes.keySet).view.map { k =>
+        val prefix = (pkgPrefixes.get(k), resolvedModulePrefixes.get(k)) match {
+          case (None, None) =>
+            throw new RuntimeException(
+              "Internal error: key in pkgPrefixes and resolvedModulePrefixes could not be found in either of them"
+            )
+          case (Some(a), None) => a.stripSuffix(".")
+          case (None, Some(b)) => b.stripSuffix(".")
+          case (Some(a), Some(b)) => s"""${a.stripSuffix(".")}.${b.stripSuffix(".")}"""
+        }
+        k -> prefix
+      }.toMap
+    resolvedPrefixes
+  }
+
+  /** Verify that no two module names collide when the given
+    * prefixes are applied.
+    */
+  def detectModuleCollisions(
+      pkgPrefixes: Map[PackageId, String],
+      interfaces: Seq[Interface],
+  ): Unit = {
+    val allModules: List[(String, PackageId)] = interfaces.view.flatMap { interface =>
+      val modules = interface.typeDecls.keys.view.map(name => name.module).toSet
+      modules.map(m =>
+        pkgPrefixes
+          .get(interface.packageId)
+          .fold(m.toString)(prefix => s"$prefix.$m") -> interface.packageId
+      )
+    }.toList
+    allModules.groupBy(_._1).foreach { case (m, grouped) =>
+      if (grouped.length > 1) {
+        val pkgIds = grouped.map(_._2)
+        throw new IllegalArgumentException(
+          s"""Duplicate module $m found in multiple packages ${pkgIds.mkString(
+            ", "
+          )}. To resolve the conflict rename the modules in one of the packages via `module-prefixes`."""
+        )
+      }
+    }
```

Comment:
```suggestion
    for ((m, grouped) <- allModules.groupBy(_._1) if grouped.length > 1) {
      val pkgIds = grouped.valuesIterator.mkString(", ")
      throw new IllegalArgumentException(
        s"Duplicate module $m found in multiple packages $pkgIds. To resolve the conflict rename the modules in one of the packages via `module-prefixes`."
      )
    }
```
Using the `for`-comprehension is probably a matter of taste but I'd recommend to not have interpolations that require curly braces in very long strings, the formatting because a bit broken and makes it less readable.

Reply:
55e100a94f1e004ffc66fd72b0a3cd6d5b18caa0 I tend to avoid for comprehensions for this type of stuff since I read @S11001001’s https://discuss.daml.com/t/pattern-always-adds-withfilter-scala-ritual/1463

Reply:
Good to know, thanks.